File: src/drawmaha_solver/rps/game.py
"""Rock-paper-scissors rules: actions, payoffs, winner determination.

Rung 0 of the validation ladder. RPS is the smallest game with a
mixed-strategy Nash equilibrium — uniform (1/3, 1/3, 1/3) — which gives the
regret-matching ledger a known exact answer to converge to before any game
tree exists. "Exploitability" here is the one-shot best-response value: what
a perfect adversary who knows the strategy (but plays the same simultaneous
game) earns per round against it.
"""

from __future__ import annotations

from enum import IntEnum

import numpy as np

# ---------------------------------------------------------------------------
# Actions and payoffs
# ---------------------------------------------------------------------------

class Action(IntEnum):
    # IntEnum (not StrEnum) so actions index payoff matrices and strategy
    # vectors directly.
    ROCK = 0
    PAPER = 1
    SCISSORS = 2

N_ACTIONS = len(Action)

# PAYOFF[a, b] = row player's chips when row plays a and column plays b:
# +1 win, 0 tie, -1 loss. Zero-sum and symmetric: PAYOFF == -PAYOFF.T.
PAYOFF = np.array(
    [
        [0.0, -1.0, 1.0],  # rock:     ties rock, loses to paper, beats scissors
        [1.0, 0.0, -1.0],  # paper:    beats rock, ties paper, loses to scissors
        [-1.0, 1.0, 0.0],  # scissors: loses to rock, beats paper, ties scissors
    ]
)

def payoff(a: Action, b: Action) -> float:
    """Player a's payoff when a meets b."""
    return float(PAYOFF[a, b])

def winner(a: Action, b: Action) -> int | None:
    """0 if action `a` wins, 1 if action `b` wins, None on a tie."""
    p = PAYOFF[a, b]
    if p > 0:
        return 0
    if p < 0:
        return 1
    return None

# ---------------------------------------------------------------------------
# Strategy metrics
# ---------------------------------------------------------------------------

def action_values(strategy: np.ndarray) -> np.ndarray:
    """Expected payoff of each pure action against a mixed strategy."""
    return PAYOFF @ np.asarray(strategy, dtype=np.float64)

def best_response_value(strategy: np.ndarray) -> float:
    """One-shot exploitability: what a best-responding adversary earns per
    round against `strategy`.

    Exactly 0 at the Nash equilibrium (uniform), positive everywhere else.
    Because RPS is symmetric zero-sum, the adversary's payoff matrix is the
    same PAYOFF, so this is just the best entry of action_values.
    """
    return float(np.max(action_values(strategy)))

File: web/rung0-viz/src/sim/game.ts
export const ROCK = 0
export const PAPER = 1
export const SCISSORS = 2
export const ACTIONS = ['rock', 'paper', 'scissors'] as const
export type Action = 0 | 1 | 2
export type Vec3 = [number, number, number]

/** PAYOFF[a][b] = row player's chips when playing a against b. Zero-sum, antisymmetric. */
export const PAYOFF: readonly (readonly number[])[] = [
  [0, -1, 1],
  [1, 0, -1],
  [-1, 1, 0],
]

/** Column PAYOFF[:, b] — what every own action would have scored against the revealed b. */
export function utilityVs(b: Action): Vec3 {
  return [PAYOFF[0][b], PAYOFF[1][b], PAYOFF[2][b]]
}

/**
 * exploit(σ) = max over b of Σₐ σ[a]·PAYOFF[b][a] — what a best-responding
 * adversary who knows σ earns per round. Exactly 0 at Nash; pure rock → 1.
 */
export function exploitability(sigma: readonly number[]): number {
  let best = -Infinity
  for (let b = 0; b < 3; b++) {
    let ev = 0
    for (let a = 0; a < 3; a++) ev += sigma[a] * PAYOFF[b][a]
    if (ev > best) best = ev
  }
  return best
}
